lib/Dialect/AIG/Transforms/GreedyCutDecomp.cpp

```diff
@@ -0,0 +1,297 @@
+          inputCutsSet.insert(cutOp);
+      }
+
+      inputCuts = std::move(inputCutsSet.takeVector());
+    }
+
+    if (inputCuts.size() <= 1)
+      return failure();
+
+    SmallVector<Value> inputs;
+    inputs.reserve(inputCuts.size());
+
+    // This is naive implementation of the cut emuration of the local inputs.
+    // This can be done more efficiently.
+
+    LLVM_DEBUG(llvm::dbgs() << "Trying to merge " << op << "\n");
+    for (unsigned i = (1 << (inputCuts.size() - 1)) - 1; i != 0; --i) {
+      auto checkSubsetMerge = [&](unsigned i) -> LogicalResult {
+        SetVector<Value> inValues;
+        llvm::SmallDenseSet<aig::CutOp, 4> cutSet;
+        SmallVector<Value> outValues;
+
+        for (unsigned j = 0; j < inputCuts.size(); ++j) {
+          if (i & (1 << j)) {
+            cutSet.insert(inputCuts[j]);
+            for (auto in : inputCuts[j].getInputs())
+              inValues.insert(in);
+            outValues.push_back(inputCuts[j].getResult(0));
+            LLVM_DEBUG(llvm::dbgs() << "Added " << inputCuts[j] << "\n");
+          }
+        }
+
+        Value singleOutput;
+        for (auto out : outValues) {
+          // Users of cuts must be closed under the cut set.
+          bool isClosed =
+              llvm::all_of(out.getUsers(), [&cutSet](Operation *user) {
+                if (auto cutOp = dyn_cast<aig::CutOp>(user))
+                  return cutSet.contains(cutOp);
+                return false;
+              });
+          inValues.remove(out);
+          if (!isClosed) {
+            if (singleOutput) {
+              LLVM_DEBUG(llvm::dbgs() << "Not closed\n");
+              return failure();
+            }
+
+            singleOutput = out;
+          }
+        }
+
+        if (!singleOutput || inValues.size() > cutLimit) {
+          LLVM_DEBUG(llvm::dbgs() << "Limit exceeded\n");
+          return failure();
+        }
+
+        SmallVector<Operation *> subsetCuts;
+        for (unsigned j = 0; j < inputCuts.size(); ++j) {
+          if (i & (1 << j))
+            subsetCuts.push_back(inputCuts[j]);
+        }
+
+        // Ok, let's merge the cuts.
+        auto cutOp = mergeCuts(op.getLoc(), subsetCuts, inValues.takeVector(),
+                               singleOutput, rewriter);
+        if (!cutOp) {
+          LLVM_DEBUG(llvm::dbgs() << "Failed to merge\n");
+          return failure();
+        }
+        return success();
+      };
+
+      // Always enable a bit for 0 (the original cut).
+      auto result = checkSubsetMerge((i << 1) | 1);
+      if (succeeded(result))
+        return result;
+    }
+    return failure();
+  }
+};
+
+struct SinkConstantPattern : public mlir::RewritePattern {
+  SinkConstantPattern(MLIRContext *context)
+      : mlir::RewritePattern(aig::CutOp::getOperationName(), 1, context) {}
+
+  LogicalResult matchAndRewrite(Operation *opT,
+                                PatternRewriter &rewriter) const override {
```

Comment:
Would `OpRewritePattern` clean up some of the `getOperationName()` and `opT -> op` casting?

Reply:
Yeah I missed OpRewritePattern for some reason ...